TypeCobol.Analysis/Cfg/ControlFlowGraphBuilder.cs

```diff
@@ -987,66 +987,177 @@ private Node ResolveProcedure(Node node, SectionNode sectionNode, SymbolReferenc
-        /// <param name="clonedPerforms">List of new PERFORMs cloned during the resolve process.</param>
-        private void ResolvePendingPERFORMProcedure(Tuple<PerformProcedure, SectionNode, BasicBlockForNodeGroup> perform, List<Tuple<PerformProcedure, SectionNode, BasicBlockForNodeGroup>> clonedPerforms)
-        {
-            PerformProcedure p = perform.Item1;
-            SectionNode sectionNode = perform.Item2;
-            BasicBlockForNodeGroup group = perform.Item3;
-
-            SymbolReference procedureReference = p.CodeElement.Procedure;
-            SymbolReference throughProcedureReference = p.CodeElement.ThroughProcedure;
-
-            Node procedureNode = ResolveProcedure(p, sectionNode, procedureReference);
-            if (procedureNode == null)
+        /// Add a Direct Call Perform to the Call Perform Relation. 
+        /// For a PERFORM THRU all intermediate procedures are added to the relation.
+        /// </summary>
+        /// <param name="caller">The caller of the PERFOM procedure</param>
+        /// <param name="callee">The PERFORM procedure called</param>
+        /// <param name="callPerformRelation">Call Perform Relation</param>
+        /// <param name="callPerformRelationDomain">The domain of the call perform relation</param>
+        private void AddDirectCallPerformRelation(PerformProcedure performCaller, Procedure caller, PerformProcedure callee, 
+            Dictionary<Procedure, Tuple<PerformProcedure, List<Node>, HashSet<Procedure>>> callPerformRelation,
+            HashSet<Procedure> callPerformRelationDomain)
+        {
+            var item = PendingPERFORMProcedures.FirstOrDefault(i => i.Item1 == callee);
+            if (item == null)
                 return;
-
-            Procedure procedure = _nodeToProcedure[procedureNode];
-            var clonedBlocksIndexMap = new Dictionary<int, int>();
-            if (throughProcedureReference != null)
+            PerformTarget performTarget = ComputePerformTarget(callee, item.Item2);
+            if (performTarget == null)
+                return;
+            foreach (Procedure calleeProcedure in performTarget.Procedures)
             {
-                Node throughProcedureNode = ResolveProcedure(p, sectionNode, throughProcedureReference);
-                if (throughProcedureNode == null)
-                    return;
+                addToRelation(calleeProcedure);
+            }
+            callPerformRelationDomain.Add(caller);
 
-                Procedure throughProcedure = _nodeToProcedure[throughProcedureNode];
-                if (procedure.Number > throughProcedure.Number)
+            void addToRelation(Procedure procedure)
+            {
+                if (!callPerformRelation.TryGetValue(caller, out var callees))
+                {
+                    callees = new Tuple<PerformProcedure, List<Node>, HashSet<Procedure>>(performCaller, new List<Node>(), new HashSet<Procedure>());
```

Comment:
`List<Node>` can be turned into `List<PerformProcedure>` if i recall correctly.

Reply:
I would like, but it's because of ControlFlowGraph<N, D>.AddRecursivePerform(PerformProcedure perform, List<N> recursiveJump)